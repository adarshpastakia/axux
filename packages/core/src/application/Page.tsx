/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { type FC } from "react";
import { Helmet } from "react-helmet-async";
import { Indicator } from "../animations";
import { AxSection } from "../components/Section";
import { AxTitle } from "../components/Title";
import { AxIcon } from "../icons/Icon";
import { type ChildrenProp, type ElementProps, type IconProp } from "../types";
import { AxErrorBoundary } from "./ErrorBoundary";

export interface PageProps extends ElementProps, IconProp, ChildrenProp {
  /**
   * page title
   */
  title?: string;
  /**
   * display title
   */
  showTitle?: boolean;
  /**
   * paper shadow
   */
  isPaper?: boolean;
  /**
   * loading state
   */
  isLoading?: boolean;
}

/**
 * A route page element providing a grid-based layout with sections such as header, footer, side panel, and content area,
 * using a grid system for organization and flexibility.
 * This layout provides responsive and visually appealing interfaces with consistent spacing and alignment.
 */
export const AxPage: FC<PageProps> = ({
  children,
  className,
  title,
  icon,
  rtlFlip,
  showTitle,
  isPaper = false,
  isLoading = false,
  ...rest
}) => {
  /** ***************** component *******************/
  return (
    <div
      {...rest}
      data-test-loading={isLoading}
      className={`ax-page ${isPaper ? "paper" : ""} ${className ?? ""}`}
    >
      <Helmet title={title} />
      {isLoading && <Indicator />}
      {showTitle && (
        <div className="ax-page__title">
          {icon && <AxIcon icon={icon} />}
          <AxTitle>{title}</AxTitle>
        </div>
      )}
      <AxSection>
        <AxErrorBoundary>{children}</AxErrorBoundary>
      </AxSection>
    </div>
  );
};
