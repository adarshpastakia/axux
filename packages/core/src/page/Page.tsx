// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { cloneElement, FC, forwardRef, ReactNodeArray, Suspense } from "react";
import { Helmet } from "react-helmet";
import { ErrorBoundary } from "../errorBoundary/ErrorBoundary";
import { AxIcon } from "../icons/Icon";
import { AxLoader } from "../loader/Loader";
import { ElementProps, IconProps, RefProp } from "../types";

/** @internal */
export interface PageProps extends IconProps, ElementProps, RefProp<HTMLDivElement> {
  /**
   * Page title
   */
  title?: string;
  /**
   * Header actions
   */
  actions?: ReactNodeArray;
  /**
   * Show loading indicator
   */
  isLoading?: boolean;
}

/**
 * Application page wrapper
 * @internal
 */
export const AxPage: FC<PageProps> = forwardRef<HTMLDivElement, PageProps>(
  ({ children, className, isLoading, icon, title, actions }, ref) => {
    return (
      <div className={`ax-page ${className ?? ""}`}>
        <Helmet title={title} />
        <div className="ax-page__header">
          {icon && (
            <div className="ax-page__icon">
              <AxIcon icon={icon} />
            </div>
          )}
          {title && <div className="ax-page__title">{title}</div>}
          {actions && (
            <div className="ax-page__actions">
              {actions.map((child: AnyObject) =>
                cloneElement(child, { type: "link", color: "primary" })
              )}
            </div>
          )}
        </div>
        <Suspense fallback={<AxLoader size="lg" color="primary" />}>
          <ErrorBoundary>
            <div className="ax-page__body" ref={ref}>
              {children}
            </div>
          </ErrorBoundary>
        </Suspense>
        {isLoading && <AxLoader size="lg" color="primary" />}
      </div>
    );
  }
);
AxPage.displayName = "AxPage";
