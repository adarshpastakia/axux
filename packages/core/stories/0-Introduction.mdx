import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

```tsx
import { AxApplicationProvider,AxViewport } from "@axux/core";
import { I18nextProvider } from "react-i18next";
import { BrowserRouter } from "react-router";

export const Application = () => {
  return (
    {/** i18n for localization **/}
    <I18nextProvider>
      {/** react router **/}
      <BrowserRouter>
        <AxApplicationProvider>
          <AxViewport>
            .
            .
            .
            .
          </AxViewport>
        </AxApplicationProvider>
      </BrowserRouter>
    </I18nextProvider>
  );
};
```
