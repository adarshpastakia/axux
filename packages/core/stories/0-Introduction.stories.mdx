import { Title, Canvas, Story } from "@storybook/addon-docs";

<Meta title="@core/Introduction" />

<Title>Core Components</Title>
