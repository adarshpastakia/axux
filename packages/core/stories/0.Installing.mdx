import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Installing" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

<cite type="code">src/index.tsx</cite>

```tsx
import { AxApplicationProvider, AxViewport } from "@axux/core";
import { Fragment, StrictMode } from "react";
import ReactDOM from "react-dom/client";
import { I18nextProvider } from "react-i18next";
import { Outlet, RouterProvider, createMemoryRouter } from "react-router";
import i18next from "./i18n";
import reportWebVitals from "./reportWebVitals";
import "./styles.css";

const root = ReactDOM.createRoot(document.getElementById("root"));

const router = createMemoryRouter([
  {
    path: "/",
    element: (
      <AxViewport>
        <div>Build the ui here</div>
        <Outlet />
      </AxViewport>
    ),
    children: [
      {
        path: "page",
        element: <Page />,
      },
      {
        path: "page",
        element: <Page />,
      },
    ],
  },
]);

const Wrapper = process.env.NODE_ENV === "development" ? Fragment : StrictMode;
root.render(
  <Wrapper>
    <AxApplicationProvider defaultPrimary="denim" defaultAccent="coral">
      <I18nextProvider i18n={i18next}>
        <RouterProvider router={router} />
      </I18nextProvider>
    </AxApplicationProvider>
  </Wrapper>,
);

// @ts-expect-error ignore
if (module.hot) {
  // @ts-expect-error ignore
  module.hot.accept();
}

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals(console.debug);
```
