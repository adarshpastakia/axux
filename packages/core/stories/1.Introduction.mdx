import {
  Title,
  Canvas,
  Story,
  Description,
  Heading,
  Subheading,
  Anchor,
} from "@storybook/addon-docs";
import { Meta } from "@storybook/blocks";
import {
  AxAnimation,
  AxApplicationProvider,
  AxAside,
  AxBreadcrumb,
  AxButton,
  AxCallout,
  AxCard,
  AxCollapsable,
  AxContent,
  AxDivider,
  AxErrorBoundary,
  AxFooter,
  AxHeader,
  AxFlexBox,
  AxLoader,
  AxIcon,
  AxMenu,
  AxMeter,
  AxSection,
  AxPage,
  AxProgress,
  AxTag,
  AxText,
  AxTheme,
  AxTitle,
  AxViewport,
  useApplicationContext,
} from "../src";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="app">
  <Subheading>Application</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxApplicationProvider</label>
    <Description of={AxApplicationProvider} />
  </div>
  <div className="comp-card">
    <label>AxViewport</label>
    <Description of={AxViewport} />
  </div>
  <div className="comp-card">
    <label>AxPage</label>
    <Description of={AxPage} />
  </div>
  <div className="comp-card">
    <label>AxErrorBoundary</label>
    <Description of={AxErrorBoundary} />
  </div>
</div>

<Anchor storyId="pageLayouts">
  <Subheading>Page Layout</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxSection</label>
    <Description of={AxSection} />
  </div>
  <div className="comp-card">
    <label>AxContent</label>
    <Description of={AxContent} />
  </div>
  <div className="comp-card">
    <label>AxAside</label>
    <Description of={AxAside} />
  </div>
  <div className="comp-card">
    <label>AxHeader</label>
    <Description of={AxHeader} />
  </div>
  <div className="comp-card">
    <label>AxFooter</label>
    <Description of={AxFooter} />
  </div>
</div>

<Anchor storyId="flexLayouts">
  <Subheading>Flex Layout</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxFlexBox</label>
    <Description of={AxFlexBox} />
  </div>
  <div className="comp-card">
    <label>AxFlexBox.Row</label>
    <Description of={AxFlexBox.Row} />
  </div>
  <div className="comp-card">
    <label>AxFlexBox.Col</label>
    <Description of={AxFlexBox.Col} />
  </div>
</div>

<Anchor storyId="typography">
  <Subheading>Typography</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxText</label>
    <Description of={AxText} />
  </div>
  <div className="comp-card">
    <label>AxText.Abbr</label>
    <Description of={AxText.Abbr} />
  </div>
  <div className="comp-card">
    <label>AxText.Mark</label>
    <Description of={AxText.Mark} />
  </div>
  <div className="comp-card">
    <label>AxText.Copy</label>
    <Description of={AxText.Copy} />
  </div>
  <div className="comp-card">
    <label>AxTitle</label>
    <Description of={AxTitle} />
  </div>
</div>

<Anchor storyId="buttons">
  <Subheading>Buttons</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxButton</label>
    <Description of={AxButton} />
  </div>
  <div className="comp-card">
    <label>AxButton.Group</label>
    <Description of={AxButton.Group} />
  </div>
  <div className="comp-card">
    <label>AxButton.Confirm</label>
    <Description of={AxButton.Confirm} />
  </div>
  <div className="comp-card">
    <label>AxButton.Action</label>
    <Description of={AxButton.Action} />
  </div>
  <div className="comp-card">
    <label>AxButton.Dropdown</label>
    <Description of={AxButton.Dropdown} />
  </div>
</div>
