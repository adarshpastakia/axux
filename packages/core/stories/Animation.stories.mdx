import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title,
} from "@storybook/addon-docs";
import { useState, useEffect } from "react";
import { AxAnimation, AxBadge, AxProgress } from "../src";

<Meta title="@core/Animations" />

<Title>Animations</Title>

<Heading storyId="core-animations--1a-check">Animated Circles</Heading>

<Canvas>
  <Story name="1a.Check">
    <AxAnimation.Check className="text-success-500" />
  </Story>
  <Story name="1b.Cross">
    <AxAnimation.Cross className="text-danger-500" />
  </Story>
  <Story name="1c.Info">
    <AxAnimation.Info className="text-primary-500" />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Check} />

<Heading storyId="core-animations--2-a-bars">Animated Indicators</Heading>

<Canvas>
  <Story name="2a.Bars">
    <AxAnimation.Bars className="text-primary-500" />
  </Story>
  <Story name="2b.Spinner">
    <AxAnimation.Spinner className="text-4xl text-accent-500" />
  </Story>
  <Story name="2c.Card">
    <AxAnimation.Card className="w-64" showIcon />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Bars} />

<Heading storyId="core-animations--3-a-progress-circle">Progress</Heading>

export const CircleTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return <AxProgress.Circle size="md" value={value} />;
};
const CircleStory = CircleTemplate.bind({});
CircleStory.args = {};

export const BarTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return (
    <div style={{ width: "40rem" }}>
      <AxProgress.Bar size="md" value={value} animate />
    </div>
  );
};
const BarStory = BarTemplate.bind({});
BarStory.args = {};

<Canvas>
  <Story name="3a.Progress Circle">{CircleStory}</Story>
  <Story name="3b.Progress Bar">{BarStory}</Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxProgress.Bar} />
