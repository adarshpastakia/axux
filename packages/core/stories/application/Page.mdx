import {
  ArgsTable,
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import {
  AxApplicationProvider,
  AxPage,
  AxCollapsable,
  AxViewport,
  useApplicationContext,
} from "../../src";
import * as PageStories from "./Page.stories";

<Meta of={PageStories} />

<Anchor storyId="page">
  <div className="invisible h-0 overflow-hidden">
    <Subheading>Page</Subheading>
  </div>
</Anchor>

# Route Page

<Description of={AxPage} />

<Primary />

<AxCollapsable>
  <div>Controls</div>
  <Controls of={PageStories.Example} />
</AxCollapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" of={AxPage} />

<hr />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { AxPage, AxHeader, AxContent } from "@axux/core";

export const RoutePage = () => {
  return (
    <AxPage isPaper isLoading={loadingState} title="Route title">
      <AxHeader>Page header controls</AxHeader>
      <AxContent>
        <p>Page content</p>
      </AxContent>
    </AxPage>
  );
};
```
