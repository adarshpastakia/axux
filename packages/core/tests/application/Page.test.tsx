/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { composeStories } from "@storybook/testing-react";
import { cleanup, render } from "@testing-library/react";
import React from "react";
import * as stories from "../../stories/application/Page.stories";

const { Example } = composeStories(stories);

describe("Page", () => {
  beforeAll(() => {
    render(
      <Example title="Page title" showTitle isLoading>
        Page content
      </Example>,
    );
  });

  it("should render", () => {
    expect(document.querySelector(".ax-page")).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
  });
  it("should render with title", () => {
    expect(document.querySelector(".ax-page__title")).not.toBeNull();
  });
  it("should render loading indicator", () => {
    expect(document.querySelector(".loading-indicator")).not.toBeNull();
  });

  afterAll(() => {
    cleanup();
  });
});
