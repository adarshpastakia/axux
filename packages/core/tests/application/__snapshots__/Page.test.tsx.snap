// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page should render 1`] = `"<div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div class="viewport-wrapper"><div data-test-loading="true" class="ax-page  "><div class="loading-indicator" aria-hidden="true"><div></div></div><div class="ax-page__title"><div class="ax-title ">Page title</div></div><div class="ax-section ">Page content</div></div></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;
